lutra/example-project/Project.prql

```diff
@@ -0,0 +1,19 @@
+
+## This module is configured to pull data from an SQLite chinook.db database
+@(lutra.sqlite {file="chinook.db"})
```

Comment:
How do we want to think about coupling the target and the query?

I would think we want to use some form of composition, where we say "associate this `chinook.db` file with this module `chinook`", but we don't couple their definitions.

At a more conceptual level, it's could be similar to applying a functors in OCaml or implementing a trait in Rust — mapping something we've created to something defined so we have a thing we can use. Another reference is dbt's approach for templating the source of a query so it's possible to run queries on a production vs dev vs personal database / schema.

I don't have a really good idea for how to implement this, and obv fine to experiment with things. But I do think that a really good design would be to make decouple the query and the target.

Reply:
Yeah, I was thinking about not having actual connection params within the PRQL source, similar to how dbt does it.

I've opted for this approach because it is way more convenient and it does not actually rule out composition - the queries are not defined within the database module. This means that one could swap out the database module and run the queries on a different database.

This could be done by defining the database module in a separate file, which is not committed to VCS or is overridden in production.

Reply:
Yes, we can think about this.

I think there's some nice design that's possible, where we map the relations to concrete tables. dbt do a nice job at the moment even if it needs a decent amount of jinja.

I'll meditate on it...